src/main/java/com/gluonhq/impl/maps/ImageRetriever.java

```diff
@@ -174,28 +180,46 @@ public void cacheImage(String url, int zoom, long i, long j) {
         }
 
         private void doCache(String urlString, int zoom, long i, long j) {
+            final URLConnection openConnection;
             try {
                 URL url = new URL(urlString);
-                try (InputStream inputStream = url.openConnection().getInputStream()) {
-                    String enc = File.separator + zoom + File.separator + i + File.separator + j + ".png";
-                    logger.info("retrieve " + urlString + " and store " + enc);
-                    File candidate = new File(cacheRoot, enc);
-                    candidate.getParentFile().mkdirs();
-                    try (FileOutputStream fos = new FileOutputStream(candidate)) {
-                        byte[] buff = new byte[4096];
-                        int len = inputStream.read(buff);
-                        while (len > 0) {
-                            fos.write(buff, 0, len);
-                            len = inputStream.read(buff);
-                        }
-                        fos.close();
-                    }
+                openConnection = url.openConnection();
+                openConnection.setConnectTimeout(TIMEOUT);
+                openConnection.setReadTimeout(TIMEOUT);
+            } catch (IOException ex) {
+                logger.log(Level.SEVERE, null, ex);
+                return;
+            }
+            InputStream inputStream = null;
+            FileOutputStream fos = null;
+            try {
```

Comment:
Using `try-with-resources` could eliminate a lot of code in `finally` block

Reply:
There is a reason for avoiding try-with-resources (GVM related).

Reply:
That explains it! :)